src/libopensc/pkcs15-pubkey.c

```diff
@@ -1499,76 +1468,127 @@ static struct ec_curve_info {
+		{"brainpoolP256r1",	"1.3.36.3.3.2.8.1.1.7",  "06092B2403030208010107", 256},
+		{"brainpoolP320r1",	"1.3.36.3.3.2.8.1.1.9",  "06092B2403030208010109", 320},
 		{"brainpoolP384r1",	"1.3.36.3.3.2.8.1.1.11", "06092B240303020801010B", 384},
 		{"brainpoolP512r1",	"1.3.36.3.3.2.8.1.1.13", "06092B240303020801010D", 512},
 
 		{"secp192k1",		"1.3.132.0.31", "06052B8104001F", 192},
 		{"secp256k1",		"1.3.132.0.10", "06052B8104000A", 256},
 
-		{"ed25519",		"1.3.6.1.4.1.11591.15.1", "06092B06010401DA470F01", 255},
+		/* OpenPGP extensions by Yubikey and GNUK are not defined in RFCs but we know the oid written to card */
+
+		{"edwards25519",	"1.3.6.1.4.1.11591.15.1", "06092B06010401DA470F01",   255},
 		{"curve25519",		"1.3.6.1.4.1.3029.1.5.1", "060A2B060104019755010501", 255},
 
+		/* RFC 8410 defined curves */
+		{"X25519",              "1.3.101.110", "06032b656e", 255},
+		{"X448",		"1.3.101.111", "06032b656f", 448},
+		{"Ed25519",             "1.3.101.112", "06032b6570", 255},
+		{"Ed448",		"1.3.101.113", "06032b6571", 448},
+		/* GnuPG openpgp curves as used in gnupg-card are equivalent to RFC8410 OIDs */
+		{"cv25519",		"1.3.101.110", "06032b656e", 255},
+		{"ed25519",		"1.3.101.112", "06032b6570", 255},
+
 		{NULL, NULL, NULL, 0}, /* Do not touch this */
 };
-
+// clang-format on
 
 int
 sc_pkcs15_fix_ec_parameters(struct sc_context *ctx, struct sc_ec_parameters *ecparams)
 {
 	int rv, ii;
+	int mapped = 0;
 
 	LOG_FUNC_CALLED(ctx);
 
 	/* In PKCS#11 EC parameters arrives in DER encoded form */
-	if (ecparams->der.value && ecparams->der.len)   {
-		for (ii=0; ec_curve_infos[ii].name; ii++)   {
-			struct sc_object_id id;
-			unsigned char *buf = NULL;
-			size_t len = 0;
-
-			sc_format_oid(&id, ec_curve_infos[ii].oid_str);
-			sc_encode_oid (ctx, &id, &buf, &len);
+	if (ecparams->der.value && ecparams->der.len && ecparams->der.len > 2) {
+		/* caller provided a der version of OID */
+		switch (ecparams->der.value[0]) {
+		case 0x06: /* OID */
+			for (ii = 0; ec_curve_infos[ii].name; ii++) {
+				struct sc_object_id id;
+				unsigned char *buf = NULL;
+				size_t len = 0;
+
+				sc_format_oid(&id, ec_curve_infos[ii].oid_str);
+				sc_encode_oid(ctx, &id, &buf, &len);
+
+				if (ecparams->der.len == len && !memcmp(ecparams->der.value, buf, len)) {
+					free(buf);
+					break; /* found ec_curve_infos[ii] */
+				}
+				free(buf); /* ii points at {NULL, NULL, NULL, 0} entry */
+			}
+			break;
 
-			if (ecparams->der.len == len && !memcmp(ecparams->der.value, buf, len))   {
-				free(buf);
+		case 0x13: /* printable string  max of 127  as per PKCS11 V 3.0 for experimental curves */
+			if (ecparams->der.value[1] != ecparams->der.len - 2) {
+				sc_log(ctx, "Unsupported ec params");
+				LOG_FUNC_RETURN(ctx, SC_ERROR_NOT_SUPPORTED);
+			}
+			for (ii = 0; ec_curve_infos[ii].name; ii++) {
+				size_t len = strlen(ec_curve_infos[ii].name);
+				if (ecparams->der.len - 2 != len || memcmp(ec_curve_infos[ii].name, ecparams->der.value + 2, len) != 0)
```

Comment:
this is printable string. Should we be strict about the case or use the `strncasecmp` to ignore the case?

Reply:
But "printable string"  was added for "PKCS11 V 3.0 for experimental curves" possibly without registered OIDs to avoid some made up unregistered OID slipping in to general use which is worse then passing in a printable string for testing experimental curves without an OID. In that case it is up to the developer to define the string and we should not second guest if it is case sensitive or not. 
 
The whole problem with 25519 curves was that to use PKCS11 required the use of an OID, and the experimenters  used OIDs registered but in private OID space, (better then made up OIDs) but not good to have standards based on private OIDs.  

So I would suggest leaving memcmp.

Reply:
Agreed. Resolving.